File: TCP_Socket/pascal/TCP_Client.pas
{$H+}   //String is an AnsiString

program TCP_Client(output);
uses
	sockets;

var
	sock_fd: longint;
	saddr: TSockAddr;
	buffer: array[0..255] of byte;
	bytes_sent, bytes_received: longint;
    tcp_message: string;
    tcp_received: string;

begin
tcp_message := 'Free Pascal Hello Server';

// Set up server address
saddr.sin_family := AF_INET;
saddr.sin_port := htons(1300); // Replace with your server's port
saddr.sin_addr := StrToNetAddr('127.0.0.1'); // Replace with your server's IP

// Create socket
sock_fd := fpSocket(AF_INET, SOCK_STREAM, 0);
if sock_fd < 0 then
begin
    writeln('Error creating socket');
    exit;
end;

// Connect to server
if fpConnect(sock_fd, @saddr, sizeof(saddr)) <> 0 then
begin
    writeln('Error connecting to server');
    CloseSocket(sock_fd);
    exit;
end;

// Send data
bytes_sent := fpSend(sock_fd, PChar(tcp_message), Length(tcp_message), 0);
if bytes_sent < 0 then
begin
    writeln('Error sending data');
end;

// Receive data
bytes_received := fpRecv(sock_fd, @buffer, sizeof(buffer), 0);
if bytes_received > 0 then
begin
    tcp_received := PChar(@buffer);
    writeln('Message Received: ', tcp_received);
end;

// Close socket
CloseSocket(sock_fd);

end.

File: TCP_Socket/c++/tcp_client.cpp
#include <iostream>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <string>

using namespace std;

int main() {
    // 1. Create a socket
    int clientSocket = socket(AF_INET, SOCK_STREAM, 0);  //AF_INET - IPv4; SOCK_STREAM - TCP
    if (clientSocket == -1) {
        cerr << "Error creating socket." << endl;
        return 1;
    }

    // 2. Define server address
    sockaddr_in serverAddress;
    serverAddress.sin_family = AF_INET;
    serverAddress.sin_port = htons(1300); // Server port
    serverAddress.sin_addr.s_addr = inet_addr("127.0.0.1"); // Server IP

    // 3. Connect to the server
    if (connect(clientSocket, (struct sockaddr*)&serverAddress, sizeof(serverAddress)) == -1) {
        cerr << "Error connecting to server." << endl;
        close(clientSocket);
        return 1;
    }
    cout << "Connected to server." << endl;

    // 4. Send data
    string message = "Hello from C++ client!";
    send(clientSocket, message.c_str(), message.length(), 0);
    cout << "Sent: " << message << endl;

    // 5. Receive data
    char buffer[1024] = {0};
    recv(clientSocket, buffer, 1024, 0);
	if (buffer[0] == 0){
		cout << "Nothing to receiver" << endl;
	}else{
	    cout << "Received from server: " << buffer << endl;
	}


    // 6. Close the socket
    close(clientSocket);
    return 0;
}
